### PR TITLE
Started typing codebase

### PR DESCRIPTION
**Pull request by WisdomPill**
I started giving types to the codebase using type hinting since python3.5 supports them and python3.5 is the minimum version required by the library.

I started with serializers and compressors.

I have some doubts about the client, specifically about key, version and prefix.

Redis supports everything that can be byte encoded like stated here https://redis.io/topics/data-types-intro

```
Redis keys are binary safe, this means that you can use any binary sequence as a key, from a string like "foo" to the content of a JPEG file. The empty string is also a valid key.
```

But other caches like `memcached` and django own `BaseCache` seem to point to key and prefix being a string and version an integer.

My question is, leaving Any instead of a more specific type or to enforce the types that I just mentioned?

By the way `django-stubs` here https://github.com/typeddjango/django-stubs/blob/master/django-stubs/core/cache/backends/base.pyi have specific types in some methods and on some not.

### CODE REVIEW

**Comment by WisdomPill** [comment]
I ended up not using str for key and prefix and int for version since it was almost mandatory from the rest of the method calls.

I am in still in doubt of using str for key and prefix, I would prefer more feedback before taking conclusions

File: django_redis/serializers/pickle.py
Comment (jdufresne) [inline django_redis/serializers/pickle.py:14]: As this code is Python 3 only, can use `super()` without arguments:

```suggestion
        super().__init__(options=options)
```

File: django_redis/client/default.py
Comment (jdufresne) [inline django_redis/client/default.py:25]: The regex is `str`, not `bytes`. Did you find it necessary to include `bytes`? If this is intended to support both, it is buggy and will produce a `BytesWarning`.

File: django_redis/client/default.py
Comment (jdufresne) [inline django_redis/client/default.py:121]: When does this function return `int`? Or `None` for that matter.

File: django_redis/serializers/pickle.py
Comment (WisdomPill) [reply django_redis/serializers/pickle.py:14]: Thanks, I did not know that

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:25]: Yeah, I fixed it but did not push the modifications yet

Comment (WisdomPill) [reply django_redis/client/default.py:25]: Sorry for that

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:121]: the delete should return an `int`, actually it returns directly what the client returns, that should be `bytes`

**Comment by WisdomPill** [comment]
I am testing locally with `mypy` I did not add it yet to the pipeline because I get a lot of errors as of now

**Comment by jdufresne** [comment]
Thanks!

I think we should hold off on landing this change until mypy passes and the type checking can be added to CI. Without that, this adds a lot of noise and no guarantees of correctness. We can leave this PR open to track efforts in that direction.

**Comment by WisdomPill** [comment]
Only one error is left.

```
django_redis/client/default.py:607: error: Item "str" of "Union[str, Any]" has no attribute "decode"
```

I cannot understand how to overcome this, it seems like the returning type is `Union[str, Any]`.

@jdufresne can you help with this?

**Comment by WisdomPill** [comment]
Another thing is about `black` and `isort` it seems like they are in contrast

about imports in line 8 of `django_redis/client/default.py`

and imports in line 1 of `django_redis/compressors/lz4.py`

**Comment by WisdomPill** [comment]
@jdufresne can you help with `black` and `isort` conflict?

**Comment by jdufresne** [comment]
Try rebasing on 3fd04a378804272d71c651db62927e7beeb4c62e. That might fix the issue.

**Comment by WisdomPill** [comment]
> Try rebasing on [3fd04a3](https://github.com/jazzband/django-redis/commit/3fd04a378804272d71c651db62927e7beeb4c62e). That might fix the issue.

Thanks it has fixed the issue.

Now the other clients are missing the type hints, I am going to work on it in the following days,
can you give a look to what's been done so far?

**Comment by jdufresne** [comment]
> Ditched relative imports in client/default.py

Was there a particular reason for this? Why can't relative imports remain?

**Comment by WisdomPill** [comment]
I find relative imports less explicit and prone to error, I am not strongly against them, but I avoid using them usually, I can restore them if necessary.

**Comment by WisdomPill** [comment]
@jdufresne I added mypy to travis, can you check everything is okay?
I also restored relative imports.

**Comment by terencehonles** [comment]
@WisdomPill did you want to update this to the latest master and I can review it after that? It would probably be good to get this into the codebase if there aren't too many errors from mypy.

**Comment by WisdomPill** [comment]
Yes sure, thanks for the help, will do that this weekend

**Comment by WisdomPill** [comment]
@terencehonles here you go, I fixed some type hints missing because of sentinel and added an ignore error for making flake8 and black compatible

**Comment by terencehonles** [comment]
Cool, thanks. I'll try to look at this, this week.

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:345]: These can probably be just inlined if they don't need this conditional checking (They don't on `master`, but not sure if that was the case before)

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:334]:
```suggestion
        client: Optional[Redis] = None,
        itersize: Optional[int] = None,
```

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:315]:
```suggestion
        client: Optional[Redis] = None,
```

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:295]:
```suggestion
        client: Optional[Redis] = None,
```

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:267]:
```suggestion
        self, key: Any, version: Optional[int] = None, client: Optional[Redis] = None
```

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:244]:
```suggestion
        client: Optional[Redis] = None,
```

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:242]:
```suggestion
        default: Optional[Any] = None,
```

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:230]:
```suggestion
        client: Optional[Redis] = None,
```

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:229]: Should these not be constrained more?

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:191]:
```suggestion
        client: Optional[Redis] = None,
```

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:131]:
```suggestion
        client: Redis = Optional[None],
```

File: django_redis/client/default.py
Comment (terencehonles) [inline django_redis/client/default.py:67]: Should this not be widened to instead?:

```suggestion
        self, write: bool = True, tried: Optional[Collection[int]] = ()
```

**Review by terencehonles — COMMENTED** [review]
I've not had a change to run `mypy` but here are some suggestions based on the diff.

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:131]: I think there is a little typo here, but I understood, it is in line with the other suggestions

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:229]: Well, yes and no. Technically key and value could be `ByteStr`, I was not sure because you could decide to use many types and they will be valid for redis. What do you suggest? I would keep it this way and discuss about it elsewhere, at the time I even asked on django stubs repository because it was not typed even there, don't know now, I will check

File: django_redis/hash_ring.py
Comment (WisdomPill) [reply django_redis/hash_ring.py:32]: Sure

File: django_redis/pool.py
Comment (WisdomPill) [reply django_redis/pool.py:9]: I remember I did try but it did not work, will try again

File: django_redis/pool.py
Comment (WisdomPill) [reply django_redis/pool.py:66]: Could you elaborate more?

File: django_redis/client/default.py
Comment (terencehonles) [reply django_redis/client/default.py:131]: lol, yeah, I should have just marked one and left you to `s/Redis/Optional[Redis]/` :sweat_smile:

File: django_redis/client/default.py
Comment (terencehonles) [reply django_redis/client/default.py:229]: I see, well I'm not sure why things are loosely defined on django-stubs, and I'm fine with leaving `Any`, but not sure if that's much better than leaving it undefined :confused:

File: django_redis/pool.py
Comment (terencehonles) [reply django_redis/pool.py:9]: I'll try to checkout the branch and run mypy myself soon.

File: django_redis/pool.py
Comment (terencehonles) [reply django_redis/pool.py:66]: Something like:

```python
from typing import Generic, TypeVar

PoolType = TypeVar('PoolType', covariant=True, bound=redis.connection.ConnectionPool)
RedisType = TypeVar('PoolType', covariant=True, bound=redis.client.Redis)

class ConnectionFactory(Generic[PoolType, RedisType]):
    def connect(self, url: str) -> RedisType: ...
    def get_or_create_connection_pool(self, params: Mapping[str, Any]) -> PoolType: ...
    def get_connection_pool(self, params: Mapping[str, Any]) -> PoolType: ...
```

which would allow someone to either write a mypy plugin or explicitly declare what types fill in the generic types. However, I'm not sure if people would be likely declaring a typed ConnectionFactory since the types will come from the options in `__init__`.

File: django_redis/pool.py
Comment (WisdomPill) [reply django_redis/pool.py:9]: fixed

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:67]: It should always be a list, if I am not mistaken I purposely changed it because before it could have been tuple and list

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:131]: I see you are a `sed` user. Nice to meet you mate

File: django_redis/compressors/lz4.py
Comment (WisdomPill) [reply django_redis/compressors/lz4.py:1]: fixed

File: django_redis/client/sentinel.py
Comment (WisdomPill) [reply django_redis/client/sentinel.py:4]: fixed

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:659]: it is already like this, am I wrong?

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:599]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:554]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:524]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:545]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:415]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:374]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:356]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:334]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:315]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:295]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:345]: I don't see it anymore

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:230]: done

File: django_redis/client/default.py
Comment (WisdomPill) [reply django_redis/client/default.py:131]: done

File: setup.cfg
Comment (terencehonles) [inline setup.cfg:54]: Do we need this? I tried removing it and it didn't seem to trigger any warnings.

File: django_redis/hash_ring.py
Comment (terencehonles) [inline django_redis/hash_ring.py:53]: This is probably the more compact form (see the last example here: https://docs.python.org/3/library/typing.html#typing.Generator)

```suggestion
    ) -> Iterator[Tuple[Optional[str], Optional[str]]]:
```

File: django_redis/hash_ring.py
Comment (terencehonles) [inline django_redis/hash_ring.py:10]: This is probably more appropriate because the order is respected but you can pass any iterable (or `Sequence[str]`).

```suggestion
        self, nodes: Iterable[str] = (), replicas: int = 128
```

File: django_redis/client/default.py
Comment (terencehonles) [reply django_redis/client/default.py:663]: I'm suggesting removing the variable and putting `glob_escape(...)` in the key_func call below.

**Review by terencehonles — APPROVED** [review]
LGTM, a few comments, but it looks like you spent some time adding more types so I'm happy with the increased coverage.

**Comment by terencehonles** [comment]
Do you think this has sufficient typing coverage to expose it to others or do you want to add more before doing so?

I added mypy to the coverage information, but I'm not sure how you can view it separately. I figured I'd let you know about the `--html-report` option and since I noticed that codecov.io used a cobertura report and I knew that mypy could generate that I figured that might make it easier to answer that question. However, I'm not sure if it really does and what the coverage report name actually does.

**Comment by terencehonles** [comment]
Looks like the documentation suggests flags are what I should have been using not the name.

**Comment by terencehonles** [comment]
So that last commit doesn't look like it was completely required. It just will show separate checks for mypy and the tests in the PR checks section. I think it's fine to leave it as it was, but what I was trying to figure out was how to see the coverage for mypy vs the tests in their UI. I had found an example repo that did what I had wanted, but it looks like the real difference was I was looking at the commit rather than the PR view. If you look at the commit view it will show the flags above the diff and you can toggle which coverage report you are looking at based on the flag that it was uploaded for.

For example, on the UI https://codecov.io/gh/jazzband/django-redis/src/c6a79d70ffa5816ac045cb26441fc6eeb0820031/django_redis/cache.py 
Screenshot from 2021-04-11 13-00-47 the two flags on the top right are toggles to pick which reports to show. I'm not sure why they don't enable that on the PR view, but just a FYI in case you were trying to look at the coverage information without generating the HTML report manually locally.

File: setup.cfg
Comment (terencehonles) [reply setup.cfg:54]: I see this is from https://black.readthedocs.io/en/stable/compatible_configs.html#id1

File: django_redis/hash_ring.py
Comment (WisdomPill) [reply django_redis/hash_ring.py:53]: But it has a `SendType` so Iterator is not really describing the `Generator`.
It is arguable that default to `None` is not a good returning type, typingwise

File: django_redis/hash_ring.py
Comment (WisdomPill) [reply django_redis/hash_ring.py:10]: I did not know `Iterable` could be used for tuple as well. I usually prefer to have a more specific type but I can't argue with simplicity

**Comment by WisdomPill** [comment]
So I can add the latest changes and then we can get this merged?

After that we can increase coverage even more with some good first issues for the young souls of the gituniverse?

File: django_redis/hash_ring.py
Comment (terencehonles) [reply django_redis/hash_ring.py:53]: But the send wouldn't be used here, right? I'm aware of the difference I just don't think it matters for this code

**Comment by WisdomPill** [comment]
@terencehonles can you give this a quick review. I think I did what was supposed to be done, but I might be wrong, the changes are too many to keep track of everything.

File: django_redis/hash_ring.py
Comment (terencehonles) [inline django_redis/hash_ring.py:17]: I don't think this is automatically implied so you might as well add it to this and the others

```suggestion
    def add_node(self, node: str) -> None:
```

**Review by terencehonles — APPROVED** [review]
I only saw a few changes since my last review so just starting from there. I noticed something that could be adjusted, but it can also come in a follow up

File: django_redis/hash_ring.py
Comment (WisdomPill) [reply django_redis/hash_ring.py]: Done

**Comment by WisdomPill** [comment]
@terencehonles friendly ping for the hopefully final review

**Comment by terencehonles** [comment]
> @terencehonles friendly ping for the hopefully final review

Lol, I was just waiting for you to merge since the first approval :joy: and the second was fine too that's why I had said the small suggestion could be made in a separate PR.

It looked good to me, but before we declare we're typed to external packages I'd like to get some more types in the system, but this is a good first step towards that.

**Comment by WisdomPill** [comment]
So for now we can merge it, not declare that the library is typed and put out some `good first issue` to finish typing?

**Comment by terencehonles** [comment]
Yeah, I'm good with that. Not sure if we _need_ to create an issue, since I don't think I would expect to have someone try to finish typing in one go. I would expect over time our coverage would get to a point where we're happy with declaring the package as typed and we'll enable that then. How we want to get to that is not _super_ important, but if you want to create an issue or issues for subparts of the project then I'm fine with that.

**Pull request merged**
